Software/Development/SemesterProject/App/dutyCycle.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * dutyCycle.c
+ *
+ *  Created on: 12 mar. 2019
+ *      Author: Aitor Teran
+ */
+//
+//###########################################################################
+
+//###########################################################################
+
+
+
+//
+// Includes
+//
+#include "Include/dutyCycle.h"
+#include "../System/Include/PWMConfig.h"
+
+
+//
+// Quasi-global variables definition
+//
+static struct dutyCycleTag
+{
+    DutyCycle legA;
+    DutyCycle legB;
+    DutyCycle legC;
+}dutyCycle;
+
+
+/* CONSIDER REVISION to reduce computing load.
+ * Function for internal use.
+ * map function performs a linear regression.
+ * Inputs are a value with its corresponding lower and upper limits
+ * as well as the lower and upper limits of the variable to return.
+ */
+Uint16 mapDuty(float x, float x_min, float x_max, Uint16 y_min, Uint16 y_max)
+{
+    return y_min + (Uint16)((float)(y_max-y_min)*(x-x_min)/(x_max-x_min));
+}
+
+/*
+ * Function for internal use.
+ * The duty cycle is checked in case it lies outside of boundaries.
+ * Later the struct containing the duty information gets updated.
+ */
+void fillDutyStruct(float duty, int leg)
+{
+
+    //Counter compare module uses the time counter variable to compare in order to set the
+    //correct PWM duty cycle, for this, it is important to know how many counts there are in a cycle.
+
+    //In the PWMConfig.c file, the period of the duty cycle is set by assigning the number of clk of a
+    //a period, this is done in the command 'EPwm1Regs.TBPRD = INTERNAL_FREQ/(2*SW_FREQ);'
+    //Since the configured carrier is a symmetrical triangular, the real period is actually twice as
+    //much but the counter will reach maximum that value.
+
+    //duty cycle float variable must then be converted into a Uint16 with correspondent
+    //input 0.00 -> output 0 and input 100.00 -> output INTERNAL_FREQ/(2*SW_FREQ).
+
+    //Error check.
+    if (duty > 100)
+    {
+        //Act accordingly.
+        //REPORT ERROR.
+        duty = 100; //For now, adjust to limit.
+    }
+    else if (duty < 0)
+    {
+        //Act accordingly.
+        //REPORT ERROR
+        duty = 0; //For now, adjust to limit.
+    }
+
+
+    //No errors so save to struct.
+    switch (leg)
+    {
+        case LEG_A:
+        {
+
+            dutyCycle.legA.dutyValue = duty;
+
+            //Transform duty to comparator.
+            dutyCycle.legA.dutyCompare = mapDuty(duty, 0.0, 100.0, 0, INTERNAL_FREQ/(2*SW_FREQ));
```

Comment:
Here the function is called with a lot of parameters that the everyday developer will not know about. Then a way of simplifying the task is to have another function which actually let's the user know what is going on:
            dutyCycle.legA.dutyCompare = mapDuty(duty)
            mapduty(duty){ // here you can include a nice explanation of what's going on
                perform_mapping(duty, 0.0, 100.0, 0, INTERNAL_FREQ/(2*SW_FREQ));}

this is just an example of the idea, the point is that the original function is plenty of stuff and we have to easy the readability.

Reply:
What about including upper and lower limits in the duty struct so that it contains all information about it?¿ then the only argument would be the struct.

Reply:
I think adding the limits to the structs is the best way of doing it